File: 3.CODIGO/Pry_TransControl/pry_transcontrol/lib/business/strategies/route_strategy.dart
import '../entities/ruta.dart';

/// Interfaz para la estrategia de cálculo de rutas
abstract class RouteStrategy {
  /// Calcula una ruta basada en la estrategia implementada
  Ruta calcularRuta(
    String origen,
    String destino,
    Coordenada coordenadaOrigen,
    Coordenada coordenadaDestino,
  );
}

/// Estrategia: Ruta más rápida
class RutaMasRapidaStrategy implements RouteStrategy {
  @override
  Ruta calcularRuta(
    String origen,
    String destino,
    Coordenada coordenadaOrigen,
    Coordenada coordenadaDestino,
  ) {
    // Simular cálculo de ruta más rápida
    const distancia = 120.0;
    const tiempoEstimado = 120; // 2 horas

    return Ruta(
      id: 'ruta_${DateTime.now().millisecondsSinceEpoch}',
      origen: origen,
      destino: destino,
      coordenadaOrigen: coordenadaOrigen,
      coordenadaDestino: coordenadaDestino,
      distanciaKm: distancia,
      tiempoEstimadoMinutos: tiempoEstimado,
      esSegura: true,
      puntosClave: [
        Coordenada(latitud: 4.7110, longitud: -74.0721), // Bogotá
        Coordenada(latitud: 5.0267, longitud: -74.0891), // Intermedio
      ],
    );
  }
}

/// Estrategia: Ruta más segura
class RutaMasSeguraStrategy implements RouteStrategy {
  @override
  Ruta calcularRuta(
    String origen,
    String destino,
    Coordenada coordenadaOrigen,
    Coordenada coordenadaDestino,
  ) {
    // Simular cálculo de ruta más segura (puede ser más larga)
    const distancia = 150.0;
    const tiempoEstimado = 180; // 3 horas

    return Ruta(
      id: 'ruta_${DateTime.now().millisecondsSinceEpoch}',
      origen: origen,
      destino: destino,
      coordenadaOrigen: coordenadaOrigen,
      coordenadaDestino: coordenadaDestino,
      distanciaKm: distancia,
      tiempoEstimadoMinutos: tiempoEstimado,
      esSegura: true,
      puntosClave: [
        Coordenada(latitud: 4.7110, longitud: -74.0721), // Bogotá
        Coordenada(latitud: 4.8090, longitud: -74.0151), // Intermedio seguro
        Coordenada(latitud: 5.0267, longitud: -74.0891), // Destino
      ],
    );
  }
}

/// Estrategia: Ruta con menor distancia
class RutaMenorDistanciaStrategy implements RouteStrategy {
  @override
  Ruta calcularRuta(
    String origen,
    String destino,
    Coordenada coordenadaOrigen,
    Coordenada coordenadaDestino,
  ) {
    // Simular cálculo de ruta con menor distancia
    const distancia = 110.0;
    const tiempoEstimado = 140; // 2:20 horas

    return Ruta(
      id: 'ruta_${DateTime.now().millisecondsSinceEpoch}',
      origen: origen,
      destino: destino,
      coordenadaOrigen: coordenadaOrigen,
      coordenadaDestino: coordenadaDestino,
      distanciaKm: distancia,
      tiempoEstimadoMinutos: tiempoEstimado,
      esSegura: false,
      puntosClave: [
        Coordenada(latitud: 4.7110, longitud: -74.0721), // Bogotá
      ],
    );
  }
}

/// Contexto para usar las estrategias
class RutaCalculadora {
  late RouteStrategy _estrategia;

  RutaCalculadora({RouteStrategy estrategia = const _DefaultRouteStrategy()})
    : _estrategia = estrategia;

  void setEstrategia(RouteStrategy estrategia) {
    _estrategia = estrategia;
  }

  Ruta calcular(
    String origen,
    String destino,
    Coordenada coordenadaOrigen,
    Coordenada coordenadaDestino,
  ) {
    return _estrategia.calcularRuta(
      origen,
      destino,
      coordenadaOrigen,
      coordenadaDestino,
    );
  }
}

/// Estrategia por defecto (ruta más rápida)
class _DefaultRouteStrategy implements RouteStrategy {
  const _DefaultRouteStrategy();

  @override
  Ruta calcularRuta(
    String origen,
    String destino,
    Coordenada coordenadaOrigen,
    Coordenada coordenadaDestino,
  ) {
    return RutaMasRapidaStrategy().calcularRuta(
      origen,
      destino,
      coordenadaOrigen,
      coordenadaDestino,
    );
  }
}

File: 3.CODIGO/TransControl/backend/src/business/strategies/route_strategy.ts
// ==========================================
// PATRÓN STRATEGY: INTERFAZ (STRATEGY)
// ==========================================
/**
 * Define el contrato común para todos los algoritmos de cálculo de rutas.
 * El patrón Strategy permite que el "Contexto" (RutaCalculadora) use esta interfaz
 * sin saber qué algoritmo específico se está utilizando.
 */
export interface RouteStrategy {
  calcularRuta(origen: string, destino: string): any;
}

// ==========================================
// PATRÓN STRATEGY: ESTRATEGIAS CONCRETAS
// ==========================================
/**
 * Cada clase "ConcreteStrategy" implementa un algoritmo distinto.
 * Al aislar estos algoritmos en clases separadas, evitamos usar bloques
 * enormes de if/else en la lógica principal y aplicamos el principio Abierto/Cerrado (OCP).
 */

export class RutaMasRapidaStrategy implements RouteStrategy {
  calcularRuta(origen: string, destino: string) {
    console.log(`Calculando la ruta MÁS RÁPIDA de ${origen} a ${destino}`);
    // Simulación de cálculo
    return {
      criterio: 'Rápida',
      tiempoEstimado: '4 horas',
      distancia: '350 km',
      peajes: 3,
      camino: `${origen} -> Autopista Principal -> ${destino}`
    };
  }
}

export class RutaMasSeguraStrategy implements RouteStrategy {
  calcularRuta(origen: string, destino: string) {
    console.log(`Calculando la ruta MÁS SEGURA de ${origen} a ${destino}`);
    // Simulación de cálculo
    return {
      criterio: 'Segura',
      tiempoEstimado: '5.5 horas',
      distancia: '400 km',
      peajes: 5,
      camino: `${origen} -> Vía Troncal Vigilada -> Destacamento Policial -> ${destino}`
    };
  }
}

export class RutaMenorDistanciaStrategy implements RouteStrategy {
  calcularRuta(origen: string, destino: string) {
    console.log(`Calculando la ruta de MENOR DISTANCIA de ${origen} a ${destino}`);
    // Simulación de cálculo
    return {
      criterio: 'Corta',
      tiempoEstimado: '4.5 horas',
      distancia: '280 km',
      peajes: 1,
      camino: `${origen} -> Carretera Secundaria Antigua -> ${destino}`
    };
  }
}

// ==========================================
// PATRÓN STRATEGY: CONTEXTO
// ==========================================
/**
 * RutaCalculadora es el "Contexto". No implementa el algoritmo de rutas por sí mismo,
 * sino que delega el trabajo al objeto "estrategia" que tiene asignado. 
 * Esto permite cambiar el algoritmo en pleno vuelo usando setEstrategia().
 */
export class RutaCalculadora {
  private estrategia: RouteStrategy;

  constructor(estrategia: RouteStrategy) {
    this.estrategia = estrategia;
  }

  setEstrategia(estrategia: RouteStrategy) {
    this.estrategia = estrategia;
  }

  ejecutarCalculo(origen: string, destino: string) {
    return this.estrategia.calcularRuta(origen, destino);
  }
}
